packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+In these situations, an alternative approach is to "soft-deprecate" the given legacy API. This is achieved by removing traces of the API from the docs, while the code keeps supporing it in the background (with or without a warning)
+
+### Components structure
+
+The contents of the `@wordpress/components` package can be divided into three groups: UI Primitives, UI Blocks, and Utilities.
+
+#### UI Primitives
+
+Lower-level components, which are usually composed together to build the second type of components: UI Blocks. UI Primitives usually don't contain translated text and aim at being highly configurable (at the expense of exposing more props and therefore more complexity).
+
+UI Primitives can be useful when in need of building an app with all new UI that looks and feels like WordPress but isn't powered by other WordPress technologies (like `@wordpress/i18n`).
+
+_[Note: we need to discuss how to properly classify/label these components. Should we introduce `primitives`, `blocks` and `utils` folders in the `src` folder? Should we include a standard way in each component's README to show which category it belongs to?]_
+
+#### UI Blocks
+
+Higher-level components, which usually offer fewer props compared to UI Primitives and contain translated text. These components use other WordPress technologies and are ready to be used in Gutenberg's UI.
+
+#### Utilities
+
+Components and other utilities function that don't necessarily render a piece of UI to screen, but are instead useful when implementing a given piece of functionality.
+
+### Components composition
+
+[To be expanded] E.g.:
+
+- Polymorphic Components
+- Using `children` vs custom render props vs arbitrary "data" props
+- Controlled and semi-controlled components
+- Composition patterns
+- Sub-components naming conventions
+- Components' layout responsibilities and boundaries (i.e., a component should only affect the layout of its children, not its own)
+- ...
+
+### APIs Consinstency
+
+[To be expanded] E.g.:
+
+- Boolean component props should be prefixed with `is*` (e.g. `isChecked`) or `has*` (`hasValue`).
+- Event callback props should be prefixed with `on*` (e.g. `onChanged`)
+- `styled` components should be suffixed with `*View` (e.g. `BoxView`)
+- ...
+
+### Technical requirements for new components
+
+The following are a set of technical requirements for all newly introduced components. These requirements are also retroactively being applied to existing components.
+
+For an example of a component that follows these requirements, take a look at [`ItemGroup`](/packages/components/src/item-group);
+
+#### TypeScript
+
+All new components should be written in TypeScript and should be typed using the `PolymorphicComponent` type (more details about polymorphism can be found above in the "Components composition" section).
```

Comment:
Should we change the name of PolymorphicComponent soon? WordPressComponent might be a good option? Given that that type can represent non-polymorphic components the name is extremely confusing, especially when used for typing existing components.

Reply:
Agreed. I will work towards renaming that type in a separate PR, then update the guidelines in this PR

Reply:
Until the package is fully typed with TS is rather difficult to enforce this rule. For now, you could word it as "we strongly encourage using TS".

Reply:
Changed the working to "We strongly encourage" and updated from `PolymorphicComponent` to `WordPressComponent` in [`aa5df68` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/aa5df68a656ef6e923107b491bcc8d31feee6c4c)